codegen/src/utils/fetch_metadata.rs

```diff
@@ -33,25 +34,81 @@ fn tokio_block_on<T, Fut: std::future::Future<Output = T>>(fut: Fut) -> T {
 
 /// Returns the metadata bytes from the provided URL.
 pub async fn fetch_metadata_bytes(url: &Uri) -> Result<Vec<u8>, FetchMetadataError> {
-    let hex = fetch_metadata_hex(url).await?;
-    let bytes = hex::decode(hex.trim_start_matches("0x"))?;
-    Ok(bytes)
-}
-
-/// Returns the raw, 0x prefixed metadata hex from the provided URL.
-pub async fn fetch_metadata_hex(url: &Uri) -> Result<String, FetchMetadataError> {
-    let hex_data = match url.scheme_str() {
+    let bytes = match url.scheme_str() {
         Some("http") | Some("https") => fetch_metadata_http(url).await,
         Some("ws") | Some("wss") => fetch_metadata_ws(url).await,
         invalid_scheme => {
             let scheme = invalid_scheme.unwrap_or("no scheme");
             Err(FetchMetadataError::InvalidScheme(scheme.to_owned()))
         }
     }?;
+
+    Ok(bytes)
+}
+
+/// Returns the raw, 0x prefixed metadata hex from the provided URL.
+pub async fn fetch_metadata_hex(url: &Uri) -> Result<String, FetchMetadataError> {
+    let bytes = fetch_metadata_bytes(url).await?;
+    let hex_data = format!("0x{}", hex::encode(bytes));
     Ok(hex_data)
 }
+enum ClientType {
+    WebSocket(jsonrpsee::async_client::Client),
+    Http(jsonrpsee::http_client::HttpClient),
+}
+
+impl ClientType {
+    fn request<'life0, 'life1, 'async_trait, R, Params>(
```

Comment:
you could also remove the `ClientType` and use `async fn request (client: impl jsonrpsee::core::client::ClientT, method, params) -> Result<R, E>`

doesn't matter

Reply:
Looks much better now thanks!